client/containers/DashboardSettings/NavBuilderRow.tsx

```diff
@@ -1,123 +1,164 @@
 	NavBuilderList?: (...args: any[]) => any;
 };
 
-const defaultProps = {
-	dropdownId: undefined,
-	NavBuilderList: undefined,
-	isStatic: false,
-	newLink: undefined,
-	updateItem: () => {},
-	removeItem: () => {},
-	index: undefined,
-};
+const NavBuilderRow = (props: Props) => {
+	const { dropdownId, NavBuilderList, index, item, newLink, isStatic = false } = props;
+	const { updateItem, pages, collections } = useContext(NavBuilderContext);
 
-type Props = OwnProps & typeof defaultProps;
+	const renderForPageOrCollection = (title: string, icon: null | string) => {
+		return (
+			<>
+				{icon && <Icon icon={icon} />}
+				{title}
+			</>
+		);
+	};
 
-const NavBuilderRow = (props: Props) => {
-	const {
-		dropdownId,
-		NavBuilderList,
-		index,
-		item,
-		pages,
-		updateItem,
-		removeItem,
-		newLink,
-		isStatic,
-	} = props;
-	const data = typeof item === 'string' ? pages.find((page) => page.id === item) : item;
-	if (!data) {
-		return null;
-	}
-	let type = 'page';
-	if (typeof data.href === 'string' && !isStatic) {
-		type = 'link';
-	}
-	if (data.children) {
-		type = 'dropdown';
-	}
+	const renderEditableTitle = (title: string) => {
+		return (
+			<InputGroup
+				small
+				placeholder="Title"
+				onChange={(evt) => {
+					updateItem(dropdownId, index, { title: evt.target.value });
+				}}
+				value={title}
+			/>
+		);
+	};
 
-	return (
-		<div className={classNames({ 'nav-builder-row-component': true, static: isStatic })}>
-			<div className="top">
-				{type === 'page' && <div className="nav-title">{data.title}</div>}
-				{type !== 'page' && (
+	const renderTop = () => {
+		// TODO(ian): Remove this branch after migration
+		if (typeof item === 'string') {
+			const page = pages.find((pg) => pg.id === item);
+			if (page) {
+				return renderForPageOrCollection(page.title, 'page-layout');
+			}
+			return null;
+		}
+		if ('type' in item) {
+			if (item.type === 'collection') {
+				const collection = collections.find((c) => c.id === item.id);
+				if (collection) {
+					const schema = getSchemaForKind(collection.kind);
+					return renderForPageOrCollection(
+						collection.title,
+						schema && schema.bpDisplayIcon,
+					);
+				}
+			} else if (item.type === 'page') {
+				const page = pages.find((pg) => pg.id === item.id);
+				if (page) {
+					return renderForPageOrCollection(page.title, 'page-layout');
+				}
+			}
+			return null;
+		}
+		if ('href' in item) {
+			return (
+				<>
+					<Icon icon="link" />
+					{renderEditableTitle(item.title)}
 					<InputGroup
 						small
-						placeholder="Title"
+						className="href-input"
+						placeholder="https://www.example.com"
 						onChange={(evt) => {
-							updateItem(dropdownId, index, { title: evt.target.value });
+							updateItem(dropdownId, index, { href: evt.target.value });
 						}}
-						value={data.title}
+						value={item.href}
 					/>
-				)}
-				{type === 'link' && (
-					<InputGroup
+				</>
+			);
+		}
+		if ('children' in item) {
```

Comment:
User defined type guards could make this a bit more readable too. Perhaps we even break out the various types of `CommunityNavigationChild` into other types, and have guards for those too. I'm a bit on the fence, though. Just thinking out loud.

Reply:
++